bots/jep/src/main/java/org/openjdk/skara/bots/jep/JEPBot.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package org.openjdk.skara.bots.jep;
+
+import org.openjdk.skara.bot.Bot;
+import org.openjdk.skara.bot.WorkItem;
+import org.openjdk.skara.forge.HostedRepository;
+import org.openjdk.skara.issuetracker.IssueProject;
+
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.List;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class JEPBot implements Bot, WorkItem {
+    final static String JEP_LABEL = "jep";
+    private final Logger log = Logger.getLogger("org.openjdk.skara.bots");
+    private final static Pattern jepMarkerPattern = Pattern.compile("<!-- jep: '(.*?)' '(.*?)' '(.*?)' -->");
+    private final HostedRepository repo;
+    private final IssueProject issueProject;
+
+    JEPBot(HostedRepository repo, IssueProject issueProject) {
+        this.repo = repo;
+        this.issueProject = issueProject;
+    }
+
+    @Override
+    public boolean concurrentWith(WorkItem other) {
+        if (!(other instanceof JEPBot otherBot)) {
+            return true;
+        }
+        return !repo.isSame(otherBot.repo);
+    }
+
+    @Override
+    public Collection<WorkItem> run(Path scratchPath) {
+        var prs = repo.pullRequests();
+        for (var pr : prs) {
+            var jepComment = pr.comments().stream()
+                    .filter(comment -> comment.author().equals(pr.repository().forge().currentUser()))
+                    .flatMap(comment -> comment.body().lines())
+                    .map(jepMarkerPattern::matcher)
+                    .filter(Matcher::find)
+                    .reduce((first, second) -> second)
+                    .orElse(null);
+            if (jepComment == null) {
+                if (pr.labelNames().contains(JEP_LABEL)) {
+                    pr.removeLabel(JEP_LABEL);
```

Comment:
There is a potential race here. The /jep command adds the label first and the comment after. Not sure how to solve that. Ideally the label should be added after the comment.

Reply:
Another case: If there is a exiting jep command comment before. We may get the wrong comment and get the wrong issue in the following code.
The error state will be fixed in the next JEPBot cycle, but it is not accepted because it seems that the error occurs frequently.

Currently, the JEPBot implements the interface `Bot` and `WorkItem`. I would like to seperate the Bot (also named JEPBot) and the WorkItem (may named JEPStateCheckWorkItem). This new WorkItem maps to the pre PR, which is similar to the PullRequestWorkItem. And we need to override the method `WorkItem#concurrentWith` carefully to let the new WorkItem not concurrent with the PullRequestWorkItem if they operate the same PR. This is a common way to solve concurrency issues in the current SKARA "framework".

Reply:
Unfortunately that won't work. The concurrentWith construct only works for WorkItems within the same bot module. An important property of the different bot modules is that they can be run independently in different Java processes. We are for instance currently not running the PullRequestBot and the CSRBot in the same JVM instance.

Reply:
> An important property of the different bot modules is that they can be run independently in different Java processes. We are for instance currently not running the PullRequestBot and the CSRBot in the same JVM instance.

From your description, I think this is a synchronization problem in a distributed system. Conceptually, the correct solution is using the distributed lock, but the current SKARA project don't have such mechanism.

Reply:
If change operations are done in the right order, it wouldn't be a problem. In this case, if the jep command added the label after the comment, it would be fine without synchronization. One can also argue that this issue will self correct anyway. The JEPBot will see the correct comment eventually and restore the label. The downside is that until that happens (which can easily take enough time to be confusing to users), the PR will show that the JEP has been targeted.

Reply:
I pondered on this some more and I think we can fix it. We can extend `CommandHandler::handle` for PRs with another parameter for returning a collection of labels to add to the PR. PullRequestCommandWorkItem calls this with an empty collection and any command that needs to add labels after the reply is posted can add them to this collection.

Reply:
Good idea. Actually, we also need to remove labels sometimes, such as the command `/jep unneeded`. What about returning a pair collections like `<labelsToAdd, labelsToRemove>` ?

Reply:
At least in the jep command case, removing before posting comment is fine. I don't think we should add a dead parameter at this time. This can easily be extended if we ever find a need for it.

Reply:
Hm, at second thought, I take that back. Removing it after adding a new comment is better.

Reply:
I updated the code just now. The new code changes the PR labels after commenting just as we discussed before.

And I refactored the method `processCommand` to avoid the multi-level nested conditional statements.

Reply:
This wasn't quite how I envisioned it. I don't think we need to overload the handle method. I also don't think we should use the return value for this. We already use the reply parameter as a kind of return value, a parameter that you modify to return a new state to the caller. My idea was to just add two more parameters to the existing handle method:
```
..., List<String> labelsToAdd, List<String> labelsToRemove)
```
The logic in `processCommand` wouldn't need to change much. It would need to instantiate two ArrayLists in the case of `isCommit==false`, and then loop through each list to process the label changes on the pr. The `pr.addComment(writer.toString())` line would need to be moved from the end to just after each of the two `handle` calls.

Reply:
I misunderstood your idea previously. It is better to add two more parameters to the existing handle method than overload it.

After adding two more parameters, the signatures of the method `handle` in all the commands need to be changed, too. Is it the expected behaviour? I need to confirm it, because it may change many files.

---

The method `CommandHandler#changeLabelsAfterComment`  can be added if we want to control it explicitly.
If the `changeLabelsAfterComment` is true, it has these three steps:
1. add/remove labels before comment (in the method `handle`)
2. comment (in the method `processCommand`)
3. add/remove labels after comment (in the method `processCommand`)

If the `changeLabelsAfterComment` is false, it has only two steps:
1. add/remove labels before comment (in the method `handle`)
2. comment (in the method `processCommand`)
 
If we choose not to add the method `CommandHandler#changeLabelsAfterComment`, we may do the test `labelsToAdd is not empty && labelsToRemove is not empty` to implicitly confirm whether the command want to change the labels after commenting. This implicit way actually make the same three steps : 
1. add/remove labels before comment (in the method `handle`)
2. comment (in the method `processCommand`)
3. add/remove labels after comment if existing (in the method `processCommand`)

What's your opinion, explicitly or implicitly?

Reply:
Now, I think the implicit way is better and follow your idea above.

Reply:
I don't think we need `CommandHandler#changeLabelsAfterComment`, it just adds extra complexity.

Reply:
Yes, I meant to change the signature. Updating it in all the command implementations is fine.